### PR TITLE
rename Generator to Coroutine

### PR DESCRIPTION
**Pull request by oli-obk**
implements https://github.com/rust-lang/compiler-team/issues/682

While I did an automated replacement, I went through all changes manually to avoid renaming things like "id generators", "code generator", ...

I renamed files where that was necessary due to the contents referring to the crate name itself (mir opt, codegen or debuginfo tests), or required by tidy (feature gate docs)

* [x] rename various remaining abbreviated references to generators.
* [x] rename files
* [x] rename folders
* [x] add renamed feature: `generators`, ...

r? @ghost

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -1485,64 +1485,64 @@ pub struct BodyId {
...
 #[derive(HashStable_Generic, Encodable, Decodable)]
-pub enum GeneratorKind {
+pub enum CoroutineKind {
     /// An explicit `async` block or the body of an async function.
-    Async(AsyncGeneratorKind),
+    Async(AsyncCoroutineKind),
 
-    /// A generator literal created via a `yield` inside a closure.
+    /// A coroutine literal created via a `yield` inside a closure.
     Gen,
```
Comment (fmease) [inline compiler/rustc_hir/src/hir.rs:1517]: s/Gen/Coroutine or sth similar?
```suggestion
    Coroutine,
```

Comment (jeffparsons) [reply compiler/rustc_hir/src/hir.rs:1517]: Fun observation: "generator" and "coroutine" both have nine letters. Does this imply that "coroutine" is equally worthy of an abbreviation, e.g., `Coro`?

I don't have a horse in this race; I'm mostly just curious about how these tradeoffs are made in rustc.

Comment (oli-obk) [reply compiler/rustc_hir/src/hir.rs:1517]: Yea that change is in my list of things to do ^^

Comment (oli-obk) [reply compiler/rustc_hir/src/hir.rs]: done

**Comment by pnkfelix** [comment]
I skimmed this. It looks good, though a line-by-line review is not feasible nor a reasonable use of effort.

(At some point one is best off trusting that the compiler is going to catch any omissions when it comes to a rewrite like this, since it seems very unlikely that an overlooked rename is going to pass the compiler unless all relevant instances in that scope were equally overlooked, which means no observable change in behavior.)

I'm assigning high priority to this because it's going to bitrot nearly immediatley.

@bors r+ rollup=never p=3

**Comment by oli-obk** [comment]
I'm rebasing and will retry the currently running PR to avoid conflicts of another PR landing.

**Comment by pthariensflame** [comment]
@oli-obk Missed a spot:  
> Coroutines, also commonly referred to as coroutines

**Pull request merged**